File: src/main/java/io/github/shaksternano/mediamanipulator/io/NamedFile.java
package io.github.shaksternano.mediamanipulator.io;

import com.google.common.io.Files;

import java.io.File;

public record NamedFile(File file, String name) {

    public NamedFile(File file) {
        this(file, file.getName());
    }

    public NamedFile(File file, String nameWithoutExtension, String extension) {
        this(file, filename(nameWithoutExtension, extension));
    }

    public String nameWithoutExtension() {
        return Files.getNameWithoutExtension(name);
    }

    public String extension() {
        return Files.getFileExtension(name);
    }

    private static String filename(String nameWithoutExtension, String extension) {
        var extensionWithDot = extension.isBlank() ? "" : "." + extension;
        return nameWithoutExtension + extensionWithDot;
    }
}

File: core/src/main/kotlin/io/github/shaksternano/borgar/core/io/NamedFile.kt
package io.github.shaksternano.borgar.core.io

import com.google.common.io.Files
import java.io.File
import java.nio.file.Path

data class NamedFile(val path: Path, val name: String) {

    val file: File
        get() = path.toFile()

    constructor(file: File) : this(file.toPath(), file.getName())

    constructor(file: File, name: String) : this(file.toPath(), name)

    constructor(file: File, nameWithoutExtension: String, extension: String) : this(
        file.toPath(), filename(nameWithoutExtension, extension)
    )

    constructor(path: Path, nameWithoutExtension: String, extension: String) : this(
        path, filename(nameWithoutExtension, extension)
    )

    fun nameWithoutExtension(): String {
        return Files.getNameWithoutExtension(name)
    }

    fun extension(): String {
        return FileUtil.getFileExtension(name)
    }
}
